File: 228 Summary Ranges/Java/Solution.java
import java.util.List;
import java.util.ArrayList;

public class Solution {
    private static void appendRange(List<String> ranges, int startNum, int endNum) {
        if (startNum == endNum) {
            ranges.add(String.format("%d", startNum));
        }
        else {
            ranges.add(String.format("%d->%d", startNum, endNum));
        }
    }

    public List<String> summaryRanges(int[] nums) {
        int len = nums.length;

        if (len == 0) {
            return new ArrayList<>();
        }

        List<String> ranges = new ArrayList<>();
        int startNum = nums[0];
        int endNum = nums[0];
        for (int i = 1; i < len; i++) {
            int currentNum = nums[i];
            int previousNum = nums[i-1];
            if (currentNum == previousNum + 1) {
                endNum = nums[i];
            }
            else {
                appendRange(ranges, startNum, endNum);
                startNum = nums[i];
                endNum = nums[i];
            }
        }

        appendRange(ranges, startNum, endNum);

        return ranges;
    }
}

File: 228 Summary Ranges/Python/Solution.py
from typing import List, Self


class Solution:
    def appendRange(self: Self, ranges: List[str], startNum: int, endNum: int) -> None:
        if startNum == endNum:
            ranges.append(f"{startNum}")
        else:
            ranges.append(f"{startNum}->{endNum}")
    
    def summaryRanges(self: Self, nums: List[int]) -> List[str]:
        l: int = len(nums)

        if l == 0:
            return []
    
        startNum: int = nums[0]
        endNum: int = nums[0]
        ranges: List[str] = []
        for i in range(1, len(nums)):
            currentNum: int = nums[i]
            previousNum: int = nums[i-1]
            if currentNum == previousNum + 1:
                endNum = currentNum
            else:
                self.appendRange(ranges, startNum, endNum)
                startNum = currentNum
                endNum = currentNum

        self.appendRange(ranges, startNum, endNum)
        
        return ranges